Add typeahead for languages that support syntax highlighting

cory2067 opened the pull request:
![image](https://cloud.githubusercontent.com/assets/8433005/25426223/718d9af8-2a3c-11e7-97d5-283e983a18c0.png)
This new feature, as suggested in #4111, is an addition to typeahead which autocompletes a programming language that has support for syntax highlighting in code blocks. Upon pushing enter, the closing backticks are also autocompleted, and the cursor is put in the middle:
![image](https://cloud.githubusercontent.com/assets/8433005/25426322/bb9e7e14-2a3c-11e7-83f6-54e67acc6f37.png)

Note that the sorting is not just alphabetical, but it attempts to arrange the options in order of popularity. The popularity ratings are generally based of the TIOBE index, with a bit of manual tweaking to improve performance. Based on `lang.json`, which contains these rankings, a Javascript file containing the dictionary of languages is constructed by a Python script and saved in `static/generated/pygment_data.js`. 

It's basically fully functional, but needs a few more additions before I'll remove the [WIP]. 
- Have not added support for `~~~` yet (currently only triple backtick)
- Have not yet added unit tests for the new typeahead

I'm particularly looking for feedback on the JavaScript generator script, because I'm not sure if there is a more "proper" way to handle it than what I'm currently doing. (I also question the quality of any code I write past 5 am)

smarx commented:
*Automated message from Dropbox CLA bot*

@cory2067, it looks like you've already signed the Dropbox CLA. Thanks!

showell commented:
In addition to `~~~`, you'll also want to add a special case for `~~~ quote`, which we don't handle through pygments, but which we provide support for.

timabbott commented on zproject/settings.py:
```diff
@@ -826,6 +826,7 @@ def get_secret(key):
             'node_modules/handlebars/dist/handlebars.runtime.js',
             'third/marked/lib/marked.js',
             'generated/emoji/emoji_codes.js',
+            'generated/pygment_data.js',
```
I think maybe this should be called `pygments_data.js`, given that `pygments` is the name of the Python library?

timabbott commented on tools/setup/lang.json:
```diff
@@ -0,0 +1 @@
+{"mql4": 9, "matlab": 33, "vbnet": 45, "fsharp": 19, "tex": 40, "go": 34, "vb.net": 45, "julia": 4, "perl": 40, "objectpascal": 42, "mql": 9, "pascal": 42, "csharp": 47, "erlang": 10, "python": 46, "lisp": 18, "scheme": 14, "rb": 39, "math": 50, "bash": 7, "r": 37, "sas": 30, "quote": 50, "abap": 27, "c++": 48, "delphi": 42, "lua": 22, "javascript": 51, "js": 43, "swift": 41, "c": 49, "java": 50, ...
```
Can you pretty-print this with one language on each line?  And maybe sort it, either by value or by language name?  Since we may be editing this by hand, we don't really want it to have a single super-long line, since diffs will be much less readable that way :)

cory2067 commented on static/js/composebox_typeahead.js:
````diff
@@ -332,18 +354,24 @@ exports.content_typeahead_selected = function (item) {
                 + '#**' + item.name + '** ');
         $(document).trigger('streamname_completed.zulip', {stream: item});
+    } else if (this.completing === 'syntax') {
+        beginning = "```" + item + "\n";
````
note to self: Fix this today, this has a side effect of deleting anything the user has already typed before/after the backticks

timabbott commented on tools/lib/provision.py:
```diff
@@ -219,6 +219,7 @@ def main(options):
     run(["sudo", "chown", "%s:%s" % (user_id, user_id), EMOJI_CACHE_PATH])
     run(["tools/setup/emoji/download-emoji-data"])
     run(["tools/setup/emoji/build_emoji"])
+    run(["tools/setup/build_pygment_data.py"])
```
Also, you should increase `PROVISION_VERSION` in `version.py`, so folks know to rerun provision after rebasing on top of this branch.

timabbott replied:
Also, need to add this to the similar section of `docs/install-generic-unix-dev.md`

timabbott commented on static/js/composebox_typeahead.js:
```diff
@@ -234,6 +241,19 @@ exports.compose_content_begins_typeahead = function (query) {
+        this.completing = 'syntax';
+        this.token = current_token.substring(3);
+        return Object.keys(pygment_data.langs);
+    }
```
We should discuss whether it makes sense to also autocomplete on ` ``` python`; GitHub supports that syntax as well so I imagine some folks will think you need a space.

timabbott commented:
I checked this out, and the experience of using it is pretty good!  I posted a bunch of small comments; looks like you also have a small linter error.

cory2067 commented:
Ok, I addressed all the things discussed above.
Triple tilde now works, and unit tests were added.

Also, I decided to allow an extra whitespace to permit stuff like ~``` py~

timabbott commented:
@cory2067 this looks great!  I noticed one bug, which is that code blocks are only valid at the start of the a line, while your highlighting logic pops up the typeahead in situations like this:
![image](https://cloud.githubusercontent.com/assets/2746074/25540921/9e4fee62-2c01-11e7-95ed-109fc729f348.png)

Your tests look like they're checking for this, but clearly they don't have the right context or something :).

Other than that, I think you can squash the commits together and write a nice commit message for the overall change, and this will be ready to merge.  Thanks @cory2067!

cory2067 commented:
Ah, I hadn't realized the code blocks needed to start at the beginning of a line. My test cases were actually testing the opposite (the tests expected the typeahead to pop up for `foo ```py`). 

Fixed that and adjusted the test cases appropriately.

timabbott commented on static/js/composebox_typeahead.js:
```diff
@@ -212,6 +217,15 @@ exports.tokenize_compose_str = function (s) {
+                if (i === 2) {
+                    return s.slice(0);
+                } else if (s[i-3] === "\n") {
+                    return s.slice(i-2);
```
Can you explain why this logic is correct?  It seems like `i-3` might underflow the array....

cory2067 replied:
True, `i-3` would be negative if you'd only typed one or two backticks. Maybe charAt(i-3) would be safer? (Because charAt is defined to just return an empty string if you specify an index that doesn't exist)

timabbott replied:
seems like you could just check the value of `i` instead?

cory2067 replied:
hm, yeah I'll just do that

timabbott commented:
Posted one comment, otherwise looks great!

timabbott commented:
Nice, merged, thanks @cory2067!  Congrats on your first big feature being merged :).
